bvm/ballerina-core/src/main/java/org/ballerinalang/BLangProgramRunner.java

```diff
@@ -69,12 +70,13 @@ public static void resumeStates(ProgramFile programFile) {
         Debugger debugger = new Debugger(programFile);
         initDebugger(programFile, debugger);
 
-        FunctionInfo functionInfo = getEntryFunctionInfo(entryPkgInfo, functionName);
+        FunctionInfo functionInfo = getMainFunctionInfo(entryPkgInfo);
         try {
             entryFuncResult = BVMExecutor.executeEntryFunction(programFile, functionInfo,
-                    extractEntryFuncArgs(functionInfo, args));
+                                                               extractEntryFuncArgs(functionInfo, args));
+            mainRunSuccessful = true;
         } finally {
-            if (!programFile.isServiceEPAvailable()) {
+            if (!mainRunSuccessful || !programFile.isServiceEPAvailable()) {
```

Comment:
Don't we need to notify exit even if the main runs successfully?

Reply:
We only need to notify exit and call `stopProgramFile` if either
- the main execution failed or
- the main succeeded AND there is no service EP available

which is what is checked in the two conditions in the `if` check.